tdrs-backend/tdpservice/parsers/parse.py

```diff
@@ -0,0 +1,131 @@
+"""Convert raw uploaded Datafile into a parsed model, and accumulate/return any errors."""
+
+
+import os
+from . import schema_defs
+from . import validators
+
+
+def parse_datafile(datafile):
+    """Parse and validate Datafile header/trailer, then select appropriate schema and parse/validate all lines."""
+    rawfile = datafile.file
+    errors = {}
+
+    document_is_valid, document_error = validators.validate_document(rawfile)
+    if not document_is_valid:
+        errors['document'] = [document_error]
+        return errors
+
+    # get header line
+    rawfile.seek(0)
+    header_line = rawfile.readline().decode().strip()
+
+    # get trailer line
+    rawfile.seek(0)
+    rawfile.seek(-2, os.SEEK_END)
+    while rawfile.read(1) != b'\n':
+        rawfile.seek(-2, os.SEEK_CUR)
+
+    trailer_line = rawfile.readline().decode().strip('\n')
+
+    # parse header, trailer
+    header, header_is_valid, header_errors = schema_defs.header.parse_and_validate(header_line)
+    if not header_is_valid:
+        errors['header'] = header_errors
+
+    trailer, trailer_is_valid, trailer_errors = schema_defs.trailer.parse_and_validate(trailer_line)
+    if not trailer_is_valid:
+        errors['trailer'] = trailer_errors
+
+    if not header_is_valid or not trailer_is_valid:
+        return errors
+
+    # determine the file type
+    schema_options = get_schema_options(header['program_type'])
+
+    # ensure file section matches upload section
+    section_names = {
+        'A': 'Active Case Data',
+        'C': 'Closed Case Data',
+        'G': 'Aggregate Data',
+        'S': 'Stratum Data',
+    }
+
+    section = header['type']
+
+    if datafile.section.find(section_names[section]) == -1:
+        errors['document'] = ['Section does not match.']
+        return errors
```

Comment:
@raftmsohani Will we need to be clever here and check for the prepended "Tribal" and "SSP" variants?

Reply:
The datafile will have the variant and since we are attaching the error to datafile, anyone who is reviewing errors should be able to tell which section it is, however, this can help if someone need to pull multiple errors and trying to filter them based on section